src/AI/ModelForCO.java

```diff
@@ -0,0 +1,74 @@
+package AI;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+
+import CommandingOfficers.Commander;
+import Engine.Army;
+import Units.UnitModel;
+
+public class ModelForCO implements Serializable
```

Comment:
I feel like I'm always having to read code in depth to figure out just what you are trying to do. Comments are great, man. Too bad this class doesn't have any.

Reply:
I thought it was clear enough from the name and constructors.
There's not much *to* it other than what amounts to boilerplate and constructors.

Reply:
But seriously, didn't we just make it so that all Commanders share the same underlying models? Why do we now need a class to keep track of models on a per-CO basis?

Reply:
The fact that there's not much to it doesn't really help, because everything needed to understand it is contextual/external. The places I see it used it just seems duplicative of the plain `UnitModel` (and there are often literally duplicate functions, to handle both it and this).
Admittedly, I haven't invested a ton of time to try and figure it out, but I also shouldn't have to reverse-engineer your logic to understand your intent. It just feels like everything you touch gets more complex for no apparent reason.

Reply:
> But seriously, didn't we just make it so that all Commanders share the same underlying models? Why do we now need a class to keep track of models on a per-CO basis?

The AIs care about the *modified* characteristics of the unit types in question, not just the *base* values.
I suppose a `UnitContext` would work as well, but it wouldn't be as obvious that we want a "type" of unit.
Wally, for example, wants to track how fast units are re: the current terrain. If he's using Sturm, Recons should look a lot more attractive.
Muriel wants to know how efficiently each COs' units can fight each other - stat modifications will definitely have implications there.

`UnitModel` used to be 100% bound up with the CO (even when it had no direct reference to the CO, its state would be modified re: any D2D/powers), and our AIs are designed around that fact.
We *could* change that but the easier solution was to make a class that represents the abstraction layer the AI wants.
> everything needed to understand it is contextual/external

A fair point, I think.
We're back to that weird place where I don't know where to put the comments to explain it, though.
> The places I see it used it just seems duplicative of the plain `UnitModel` (and there are often literally duplicate functions, to handle both it and this).

I think this is address sufficiently above?
> I also shouldn't have to reverse-engineer your logic to understand your intent

Some minimal amount of that is necessary, lest we start needing to comment... IDK, basic for-loop headers.
You're probably right that some stuff is more obtuse than it should be, though.
> It just feels like everything you touch gets more complex for no apparent reason.

It's getting more complex because I would like to handle more complex scenarios gracefully.
Perhaps I'm optimizing for the edge cases a bit too much, because it's the most "interesting" to me.

Reply:
> The AIs care about the modified characteristics of the unit types in question, not just the base values.

Yeah, sure. But this doesn't allow that. It literally just maps COs to their UnitModels. The models themselves aren't differentiated/modified per the CO's abilities (since we no longer have a way to do that outside of UnitContext).

It's also only used in Muriel, and only to separate enemy unit types into lists by CO. Wally doesn't even use it (It only invokes the ModelForCO type because CPI now returns it, but then it just pulls out the UnitModel itself).

Am I just grossly missing something, or is this class just a more complicated way of doing nothing special?

> EDIT: I've thought a bit more about it, and I think I've hit upon our methodology difference:
My instinct, when adding functionality, is to make a general infrastructure to support permutations of that form.
Yours is to change code outside of your new addition as little as possible

I think both values are important: avoid changing what doesn't need to be changed (because needless code churn is bad), and also make changes of the same type easier in the future (because needless code churn is bad). ModelForCO... doesn't do anything but add complexity, as far as I can tell.

Reply:
> Yeah, sure. But this doesn't allow that. It literally just maps COs to their UnitModels.

Having CO + UnitModel lets you calculate whatever it is you want. I don't see the problem?
> The models themselves aren't differentiated/modified per the CO's abilities (since we no longer have a way to do that outside of UnitContext).
> 
> It's also only used in Muriel, and only to separate enemy unit types into lists by CO.

I guess another way to do it would be to use `UnitContext`s, but then you can't as easily use them as keys for stuff like the `UnitMatchupAndMetaInfo`.
You can just use `UnitModel` itself, but then you can't store different matchups per CO - and that's something you specifically called out as being broken (I had the code commented out).
I could revert that change so Muriel ends up entirely CO-agnostic as far as unit builds, but that's sort of a bad idea when people like Max and Grit exist.
> Wally doesn't even use it (It only invokes the ModelForCO type because CPI now returns it, but then it just pulls out the UnitModel itself).

Correct. I didn't feel like putting a bunch of effort in to fix his current state when he needs more involved surgery anyway.
Wally is broken by this change, but not badly (given our current CO loadout), and I do intend to fix that later along with more involved changes.
It just seemed kinda rude to break the AI you wrote and not fix it.
> Am I just grossly missing something, or is this class just a more complicated way of doing nothing special?

If you know of a datastructure that can take a CO and a `UnitModel` as a duet unique key and you want to make the AIs start using that, I guess I'm down to try doing it.
Otherwise, all I can think of is a two-tier `HashMap`... and I feel like that will have bigger impacts on the clarity of the code than this does.
> I think both values are important: avoid changing what doesn't need to be changed (because needless code churn is bad), and also make changes of the same type easier in the future (because needless code churn is bad).

Quite fair.
> ModelForCO... doesn't do anything but add complexity, as far as I can tell.

It adds a new object type to be a key into Maps that want to track some value for a unit type for a specific CO.
`UnitContext` does not work since there's no sensible way to set up `.Equals()` such that it would do what you expect in both this and its other use cases (I guess you could make it ignore `UnitState` values if there's no individual unit reference, but that seems like a great way to get subtle bugs...).
`UnitModel` does not work since you entirely lose the CO information if you just use it.

Reply:
> Having CO + UnitModel lets you calculate whatever it is you want. I don't see the problem?

This is the kind of raison d'être that should probably go in the class javadoc. That said, you are now creating ephemeral units every time you need CO-specific unit info (e.g. to calculate matchups). Seems like it would be cheaper to just create one unit per type per CO up front instead of one unit per type per CO per other type per other CO ephemerally. But that feels almost like going back to CO-specific unit models, except they'd be mutable.

> It just seemed kinda rude to break the AI you wrote and not fix it.

Hah, well I do appreciate that.

Reply:
> That said, you are now creating ephemeral units every time you need CO-specific unit info (e.g. to calculate matchups).

I thought Muriel was already doing that?
`ModelForCO` is just a pair of final pointers, so it's not a full unit.
> Seems like it would be cheaper to just create one unit per type per CO up front instead of one unit per type per CO per other type per other CO ephemerally. But that feels almost like going back to CO-specific unit models, except they'd be mutable.

I mean, caching something like that in your AI isn't the same as CO-specific unit models.
If you really wanted to be penny-pinching about it, I'd say to allocate exactly two `UnitContext`s (for attacker/defender) and swap out properties as you do calcs.
> *Edit: I see you did add a javadoc note that could be construed as providing that insight, but it's pretty oblique; it describes what an AI could use it to do, but not how this class helps it to do that. Or maybe I was just too stupid or distracted to make the necessary connection, who knows.

I thought "track stuff" plus "overrides `equals()` and `hashCode()`" was enough to be reasonably clear.
Should I add "Overrides equals() and hashCode() so it can be used as a Map key."?
I do want the docs to be useful, and part of that is keeping them minimal enough that people will actually read them, lol.

Reply:
It's more the "colocates CO and UnitModel to allow creating a unit to calculate CO-specific unit stats that aren't directly available from just the model", and less the "now you can use this pair as a map key" rationale I'm looking for to explain the class's existence, and that is only indirectly mentioned.

Reply:
Oh. Okay.